### PR TITLE
6 a

### PR DESCRIPTION
**Pull request by AngelRRuizM**
Tests are going to fail, still missing to refactor them and chenage the seeders

### CODE REVIEW

File: server/middleware/validations/skill-validation.js [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+
+let ErrorFactory = require('../../resources').ErrorFactory;
+let Errors = require('../../resources').Errors;
+let topicFields = require('../../resources').fields;
+
+const validateSkill = (req, res, next) => {
+    let skill = req.body;
+
+    let requiredFields = [
+        'name',
```
Comment (Syul968) [inline server/middleware/validations/skill-validation.js:11]: Will we need to update our data model to include `name`?

File: server/middleware/validations/skill-validation.js [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+    // Required fields
+    for(let field of requiredFields) {
+        if(req.method == 'POST' && !skill[field]) {
+            let error = ErrorFactory.buildError(Errors.MISSING_FIELD, field);
+            return res.status(error.status).send({ error: error });
+        }
+    }
+
+    // Name min length 2 characters
+    if(skill.name && skill.name <= 1) {
```
Comment (Syul968) [inline server/middleware/validations/skill-validation.js]: `.length`?

File: server/middleware/validations/skill-validation.js [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+        let error = ErrorFactory.buildError(Errors.SHORT_STRING, 'name', '1');
+        return res.status(error.status).send({ error: error });
+    }
+
+    // Name min length 2 characters
```
Comment (Syul968) [inline server/middleware/validations/skill-validation.js]: nit: what is this validating?

File: server/middleware/validations/skill-validation.js [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+    ];
+    
+    // Should have skill
+    if(Object.keys(skill).length === 0) {
+        let error = ErrorFactory.buildError(Errors.MISSING_FIELD, 'skill (object)');
+        return res.status(error.status).send({ error: error });
+    }
+
+    // Required fields
+    for(let field of requiredFields) {
```
Comment (Syul968) [inline server/middleware/validations/skill-validation.js:23]: Why do we need to provide `field` with skill? isn't it part of topic, which we are referencing by id?

File: server/models/user.js [hunk 5]
```diff
@@ -38,8 +34,14 @@ const userSchema = new mongoose.Schema({
...
             skills: {
                 type: [{
-                    placeHolder: {
-                        type: String
+                    topic: {
+                        type: mongoose.Schema.Types.ObjectId,
+                        ref: 'Topic',
+                        required: true
+                    },
+                    exeprience:{
```
Comment (Syul968) [inline server/models/user.js]: nit: spelling

File: server/controllers/skill.js [hunk 6]
```diff
@@ -0,0 +1,430 @@
...
+
+        // Validate tutor exists
+        let tutor = await User.findById(tutorId).exec();
+        if(!tutor || !tutor.tutorDetails ) {
+            let error = ErrorFactory.buildError(Errors.OBJECT_NOT_FOUND, 'tutor');
+            return res.status(error.status).send({ error: error });
+        }
+
+        // Validate skill exists
+        let nSkill = req.body;
```
Comment (Syul968) [inline server/controllers/skill.js:55]: isn't this validated by middleware?

File: server/controllers/skill.js [hunk 7]
```diff
@@ -0,0 +1,430 @@
...
+
+        let session = await mongoose.startSession();
+
+        await session.startTransaction();
+
+        let relatedTopic = await Topic.findOne({'Name': nSkill.name, 'Field': nSkill.Field}).exec();
+
+        //console.log(relatedTopic);
```
Comment (Syul968) [inline server/controllers/skill.js]: nit: will we need this?

File: server/controllers/skill.js [hunk 8]
```diff
@@ -0,0 +1,430 @@
...
+                experience: nSkill.experience,
+                _id: skillId
+            });
+
+            if(!relatedTopic.Tutors)
+                relatedTopic.Tutors = [];
+
+            relatedTopic.Tutors.push(tutorId);
+
+            console.log(tutor);
```
Comment (Syul968) [inline server/controllers/skill.js]: Do we need this?

File: server/controllers/skill.js [hunk 9]
```diff
@@ -0,0 +1,430 @@
...
+                    });
+                })
+                .catch(async (err) => {
+                    await session.abortTransaction();
+                    return res.status(500).send({error: ErrorFactory.buildError(Errors.DATABASE_ERROR, err)});
+                })
+            })
+            .catch(async (err) => {
+                await session.abortTransaction();
+                return res.status(500).send({error: ErrorFactory.buildError(Errors.DATABASE_ERROR, err)});
```
Comment (Syul968) [inline server/controllers/skill.js:103]: we can build the error first and avoid hardcoding status.

**Review by Syul968 — APPROVED** [review]
That's a huge contribution, thank you! 👍
I'd only suggest to check the name length validation.

**Review by RComonfort — APPROVED** [review]
Looks good!

**Pull request merged**